akka-projection-eventsourced/src/main/scala/akka/projection/eventsourced/EventEnvelope.scala

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.projection.eventsourced
+
+import akka.persistence.query.Offset
+import akka.util.HashCode
+
+object EventEnvelope {
+  private[akka] def apply[Event](eventEnvelope: akka.persistence.query.EventEnvelope): EventEnvelope[Event] = {
+    new EventEnvelope(
+      offset = eventEnvelope.offset,
+      persistenceId = eventEnvelope.persistenceId,
+      sequenceNr = eventEnvelope.sequenceNr,
+      event = eventEnvelope.event.asInstanceOf[Event],
+      timestamp = eventEnvelope.timestamp)
+  }
+
+  def apply[Event](
+      offset: Offset,
+      persistenceId: String,
+      sequenceNr: Long,
+      event: Event,
+      timestamp: Long): EventEnvelope[Event] =
+    new EventEnvelope(offset, persistenceId, sequenceNr, event, timestamp)
+
+  def unapply[Event](arg: EventEnvelope[Event]): Option[(Offset, String, Long, Event, Long)] =
+    Some((arg.offset, arg.persistenceId, arg.sequenceNr, arg.event, arg.timestamp))
+}
+
+final class EventEnvelope[Event](
+    val offset: Offset,
+    val persistenceId: String,
+    val sequenceNr: Long,
+    val event: Event,
+    val timestamp: Long) {
+
+  override def hashCode(): Int = {
+    var result = HashCode.SEED
+    result = HashCode.hash(result, offset)
+    result = HashCode.hash(result, persistenceId)
+    result = HashCode.hash(result, sequenceNr)
+    result = HashCode.hash(result, event)
+    result = HashCode.hash(result, timestamp)
+    result
+  }
+
+  override def equals(other: Any): Boolean =
+    other match {
+      case that: EventEnvelope[_] =>
+        offset == that.offset &&
+        persistenceId == that.persistenceId &&
+        sequenceNr == that.sequenceNr &&
+        event == that.event &&
+        timestamp == that.timestamp
+
+      case _ => false
+    }
+
+  def copy(
```

Comment:
This would probably cause binary compatibility trouble in future if we add anything (default values are evil).
Skip this, and if we see a need for amending we should add `withX` methods.